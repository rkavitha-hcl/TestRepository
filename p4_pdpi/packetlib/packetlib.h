#ifndef GOOGLE_P4_PDPI_PACKETLIB_PACKETLIB_H_
#define GOOGLE_P4_PDPI_PACKETLIB_PACKETLIB_H_

#include "absl/status/statusor.h"
#include "absl/strings/string_view.h"
#include "p4_pdpi/packetlib/packetlib.pb.h"

namespace packetlib {

// Parses the given packet. Parsing is a total function, and any aspect that
// cannot be parsed correctly will be put into `payload` of `Packet`.
//
// Even invalid packets will be parsed into the header structure of `Packet`
// when possible. For instance an invalid checksum will be parsed. However, not
// all invalid packets can be parsed into the header structure. Specifically, if
// trying to represent the packet would lose information, the function will
// instead not parse that header and put the data in `payload` instead. For
// example IPv4 packets with options is treated this way (since the library does
// not support options and thus has no place in `Ipv4Header` for options).
//
// Parsing starts with the given header (defaulting to Ethernet).
//
// Guarantees for `packet = ParsePacket(data)`:
// 1. Valid packets are valid:
//    `packet.reasons_invalid.empty()` implies `ValidatePacket(packet).ok()`.
// 2. Parsing is loss-less:
//    Running `serialized = RawSerializePacket(packet);` guarantees
//    `serialized.ok() && *serialized == data`.
// 3. If a header is supported by packetlib, it will be parsed. Partially
//    supported headers may not be parsed, but then `reason_unsupported`
//    will indicate what unsupported feature the packet uses, and the
//    unsupported header will appear uninterpreted in the payload.
Packet ParsePacket(absl::string_view input,
                   Header::HeaderCase first_header = Header::kEthernetHeader);

// Validates packets by checking that:
// 1. Headers appear in a valid order, and that fields indicating the next
//    header match the actual next header for all supported header types.
// 2. Each field is specified, and of the correct format.
// 3. That computed fields have the right value.
// 4. The packet has the required minimum size.
absl::Status ValidatePacket(const Packet& packet);

// Same as ValidatePacket, but returns a list of reasons why the packet isn't
// valid instead.
std::vector<std::string> PacketInvalidReasons(const Packet& packet);

// Seralizes a given packet. The packet may miss computed fields, which will be
// filled in automatically when missing (but not changed if they are present).
// Serialization succeeds iff `ValidatePacket(packet).ok()` after
// calling `UpdateComputedFields(packet)`; an error status is returned
// otherwise.
absl::StatusOr<std::string> SerializePacket(Packet packet);

// Seralizes a given packet without checking header invariants. All fields must
// be present and use a valid value (according to ir::Format), but otherwise no
// requirements are made on the set of headers; they will just be serialized in
// order without checking, if computed fields are correct, header order is
// valid, etc.
absl::StatusOr<std::string> RawSerializePacket(const Packet& packet);

// Updates all computed fields that are missing. Computed fields that are
// already present are not modified. Returns true iff any changes where made.
// Fails if fields that are required for determining computed fields are missing
// or invalid.
absl::StatusOr<bool> UpdateComputedFields(Packet& packet);

// Returns the size of the given packet in bits, starting at the nth header and
// ignoring all headers before that. Works even when computed fields are
// missing.
// Returns error if
// - `start_header_index` is not in [0, packet.headers().size()], or
// - `packet.headers(i)` is uninitialized for i in
//   [start_header_index, packet.headers().size()]
absl::StatusOr<int> PacketSizeInBits(const Packet& packet,
                                     int start_header_index = 0);

// Like `PacketSizeInBits`, but returns size in bytes, or an error if the bit
// size is not divisible by 8.
absl::StatusOr<int> PacketSizeInBytes(const Packet& packet,
                                      int start_header_index = 0);

// Computes the checksum of an IPv4 header. All fields must be set and valid
// except possibly the checksum, which is ignored.
absl::StatusOr<uint16_t> Ipv4HeaderChecksum(const Ipv4Header& header);

std::string HeaderCaseName(Header::HeaderCase header_case);

}  // namespace packetlib

#endif  // GOOGLE_P4_PDPI_PACKETLIB_PACKETLIB_H_
