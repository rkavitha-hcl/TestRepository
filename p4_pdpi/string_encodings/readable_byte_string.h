#ifndef GOOGLE_P4_PDPI_PACKETLIB_READABLE_BIT_STRING_H_
#define GOOGLE_P4_PDPI_PACKETLIB_READABLE_BIT_STRING_H_

#include "absl/status/statusor.h"

namespace pdpi {

// Library to write down byte strings in a readable manner. This is useful e.g.
// for writing down network packets in a readable manner.
//
// Example:
// R"PB(
//   # ethernet header
//   ethernet_source: 0x112233445566
//   ethernet_destination: 0xaabbccddeeff
//   ether_type: 0x0800
//   # IPv4 header:
//   version: 0x4
//   ihl: 0x5
//   dhcp: 0b011011)PB"
//
// Supports comments (using #), annotations of what a group of bits represents
// (string before the colon), hex strings, base-2 strings.
absl::StatusOr<std::string> ReadableByteStringToByteString(
    absl::string_view readable_byte_string);

}  // namespace pdpi

#endif  // GOOGLE_P4_PDPI_PACKETLIB_READABLE_BIT_STRING_H_
