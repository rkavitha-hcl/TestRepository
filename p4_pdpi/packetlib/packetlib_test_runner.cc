#include <iostream>
#include <string>
#include <utility>
#include <vector>

#include "absl/strings/ascii.h"
#include "absl/strings/escaping.h"
#include "google/protobuf/util/message_differencer.h"
#include "gutil/testing.h"
#include "p4_pdpi/packetlib/packetlib.h"
#include "p4_pdpi/string_encodings/readable_byte_string.h"

namespace packetlib {

constexpr char kBanner[] =
    "=========================================================================="
    "======\n";
constexpr char kInputHeader[] =
    "-- INPUT -----------------------------------------------------------------"
    "------\n";
constexpr char kOutputHeader[] =
    "-- OUTPUT ----------------------------------------------------------------"
    "------\n";
constexpr char kRoundtripHeader[] =
    "-- ROUNDTRIP ERRORS ------------------------------------------------------"
    "------\n";

// Parses packet from bytes, and if it succeeds, re-serializes the parsed packet
// to check that the resulting bytes match the original input.
void RunPacketParseTest(const std::string& name,
                        const std::string& readable_byte_string) {
  std::cout << kBanner << "Parsing test: " << name << "\n" << kBanner;
  std::cout << kInputHeader << absl::StripAsciiWhitespace(readable_byte_string)
            << "\n"
            << kOutputHeader;

  // Attempt to parse.
  auto byte_string = pdpi::ReadableByteStringToByteString(readable_byte_string);
  if (!byte_string.ok()) {
    std::cout << "TEST BUG, DO NOT"
              << " SUBMIT! ReadableByteStringToByteString failed: "
              << byte_string.status() << "\n";
    return;
  }
  Packet packet = ParsePacket(*byte_string);
  std::cout << packet.DebugString() << "\n";

  // Check roundtrip if parsing succeeded.
  if (!packet.reasons_invalid().empty()) return;
  auto byte_string_after_roundtrip = SerializePacket(packet);
  if (!byte_string_after_roundtrip.ok()) {
    std::cout << kRoundtripHeader << byte_string_after_roundtrip.status()
              << "\n";
  } else if (*byte_string_after_roundtrip != *byte_string) {
    std::cout << kRoundtripHeader
              << "Original packet does not match packet after parsing and "
              << "reserialization.\nOriginal packet:\n"
              << absl::BytesToHexString(*byte_string)
              << "\nParsed and reserialized packet:\n"
              << absl::BytesToHexString(*byte_string_after_roundtrip) << "\n";
  }
}

// Validates the packet, and if it's not valid, attempts to
// UpdateComputedFields and revalidate.
// Then attempt to serialize the packet, and if this succeeds, parse the
// serialized packet back and verify that it matches the original packet.
void RunProtoPacketTest(const std::string& name, Packet packet) {
  std::cout << kBanner << "Proto packet test: " << name << "\n" << kBanner;
  std::cout << kInputHeader << "packet =" << std::endl
            << packet.DebugString() << std::endl
            << kOutputHeader;

  auto valid = ValidatePacket(packet);
  std::cout << "ValidatePacket(packet) = " << valid << std::endl;

  if (!valid.ok()) {
    std::cout << std::endl << "UpdateComputedFields(packet) = ";
    if (auto updated = UpdateComputedFields(packet); updated.ok()) {
      std::cout << (*updated ? "true" : "false") << std::endl;
      if (*updated) {
        std::cout << "packet =" << std::endl
                  << packet.DebugString() << std::endl;
        // Try validating once more.
        std::cout << "ValidatePacket(packet) = " << ValidatePacket(packet)
                  << std::endl;
      }
    } else {
      std::cout << updated.status() << std::endl;
    }
  }

  // Try serializing (valid or invalid) packet.
  absl::StatusOr<std::string> bytes = SerializePacket(packet);
  std::cout << "Serialize(Packet) = " << bytes.status() << "\n\n";
  if (!bytes.ok()) return;

  // Test if the packet can be parsed back.
  auto reparsed_packet = ParsePacket(*bytes);
  google::protobuf::util::MessageDifferencer differ;
  std::string diff;
  differ.ReportDifferencesToString(&diff);
  if (!differ.Compare(packet, reparsed_packet)) {
    std::cout << kRoundtripHeader
              << "Original packet does not match packet after serialization "
                 "and reparsing:\n"
              << diff << "\n\n";
  }
}

void main() {
  RunPacketParseTest("Ethernet packet (valid)", R"PB(
    # ethernet header
    ethernet_source: 0x112233445566
    ethernet_destination: 0xaabbccddeeff
    ether_type: 0x0001  # This means size(payload) = 1 byte.
    # payload
    payload: 0x01
  )PB");
  RunPacketParseTest("Ethernet packet (invalid)", R"PB(
    # ethernet header
    ethernet_source: 0x112233445566
    ethernet_destination: 0xaabbccddeeff
    ether_type: 0x0001  # This means size(payload) = 1 byte.
    # payload
    pqyload: 0x0102  # 2 bytes, but ether_type says 1 byte.
  )PB");
  RunPacketParseTest("IPv4 packet (invalid)", R"PB(
    # ethernet header
    ethernet_source: 0x112233445566
    ethernet_destination: 0xaabbccddeeff
    ether_type: 0x0800
    # IPv4 header:
    version: 0x4
    ihl: 0x5
    dhcp: 0b011011
    ecn: 0b01
    total_length: 0x6fc6
    identification: 0xa3cd
    flags: 0b000
    fragment_offset: 0b0000000000000
    ttl: 0x10
    protocol: 0x05  # some unsupported protocol
    checksum: 0x1234
    ipv4_source: 0x0a000001
    ipv4_destination: 0x14000003
    # other headers:
    payload: 0x1234
  )PB");
  RunPacketParseTest("IPv4 packet (valid)", R"PB(
    # ethernet header
    ethernet_source: 0x112233445566
    ethernet_destination: 0xaabbccddeeff
    ether_type: 0x0800
    # IPv4 header:
    version: 0x4
    ihl: 0x5
    dhcp: 0b011011
    ecn: 0b01
    total_length: 0x0016
    identification: 0xa3cd
    flags: 0b000
    fragment_offset: 0b0000000000000
    ttl: 0x10
    protocol: 0x05  # some unsupported protocol
    checksum: 0xb2e7
    ipv4_source: 0x0a000001
    ipv4_destination: 0x14000003
    # other headers:
    payload: 0x1234
  )PB");
  RunPacketParseTest("IPv6 packet (invalid)", R"PB(
    # ethernet header
    ethernet_source: 0x554433221100
    ethernet_destination: 0xffeeddccbbaa
    ether_type: 0x86DD
    # IPv6 header:
    version: 0x4
    dhcp: 0b011011
    ecn: 0b01
    flow_label: 0x12345
    payload_length: 0x0000
    next_header: 0x90  # some unassigned protocol
    hop_limit: 0xff
    ipv6_source: 0x00001111222233334444555566667777
    ipv6_destination: 0x88889999aaaabbbbccccddddeeeeffff
    # other headers:
    payload: 0x12
  )PB");
  RunPacketParseTest("IPv6 packet (valid)", R"PB(
    # ethernet header
    ethernet_source: 0x554433221100
    ethernet_destination: 0xffeeddccbbaa
    ether_type: 0x86DD
    # IPv6 header:
    version: 0x6
    dhcp: 0b011011
    ecn: 0b01
    flow_label: 0x12345
    payload_length: 0x0001
    next_header: 0x90  # some unassigned protocol
    hop_limit: 0x03
    ipv6_source: 0x00001111222233334444555566667777
    ipv6_destination: 0x88889999aaaabbbbccccddddeeeeffff
    # other headers:
    payload: 0x12
  )PB");

  RunProtoPacketTest("IPv4 without computed fields",
                     gutil::ParseProtoOrDie<Packet>(R"PB(
                       headers {
                         ethernet_header {
                           ethernet_source: "11:22:33:44:55:66"
                           ethernet_destination: "aa:bb:cc:dd:ee:ff"
                           ethertype: "0x0800"
                         }
                       }
                       headers {
                         ipv4_header {
                           ihl: "0x5"
                           dscp: "0x1b"
                           ecn: "0x1"
                           identification: "0xa3cd"
                           flags: "0x0"
                           fragment_offset: "0x0000"
                           ttl: "0x10"
                           protocol: "0x05"
                           ipv4_source: "10.0.0.1"
                           ipv4_destination: "20.0.0.3"
                         }
                       }
                       payload: "0xabcd"
                     )PB"));

  RunProtoPacketTest("IPv4 with various invalid fields",
                     gutil::ParseProtoOrDie<Packet>(R"PB(
                       headers {
                         ethernet_header {
                           ethernet_destination: "aa:bb:cc:dd:ee:ff"
                           ethertype: "0x0800"
                         }
                       }
                       headers {
                         ipv4_header {
                           version: "0x3"
                           ihl: "0x5k"
                           dscp: "0x1b"
                           ecn: "0x1"
                           identification: "0xa3cd"
                           flags: "0x0"
                           fragment_offset: "0x0000"
                           ttl: "0x10"
                           protocol: "0x05"
                           ipv4_source: "ffff:1::"
                           ipv4_destination: "20.0.0.3"
                         }
                       }
                       payload: "0xabcd"
                     )PB"));

  RunProtoPacketTest("IPv6 without computed fields",
                     gutil::ParseProtoOrDie<Packet>(R"PB(
                       headers {
                         ethernet_header {
                           ethernet_source: "11:22:33:44:55:66"
                           ethernet_destination: "aa:bb:cc:dd:ee:ff"
                           ethertype: "0x86dd"
                         }
                       }
                       headers {
                         ipv6_header {
                           dscp: "0x1b"
                           ecn: "0x1"
                           flow_label: "0x12345"
                           next_header: "0x05"
                           hop_limit: "0x10"
                           ipv6_source: "::"
                           ipv6_destination: "f::f"
                         }
                       }
                       payload: "0xabcd"
                     )PB"));

  RunProtoPacketTest("IPv6 with various invalid fields",
                     gutil::ParseProtoOrDie<Packet>(R"PB(
                       headers {
                         ethernet_header {
                           ethernet_source: "11:22:33:44:55:66"
                           ethernet_destination: "aa:bb:cc:dd:ee:ff"
                           ethertype: "0x86dd"
                         }
                       }
                       headers {
                         ipv6_header {
                           version: "0x4"
                           dscp: "1b"
                           ecn: "0b01"
                           flow_label: "0x1234"
                           payload_length: "0x0000"
                           next_header: "0x050"
                           hop_limit: "0x1"
                           ipv6_source: "20.0.0.3"
                           ipv6_destination: ":"
                         }
                       }
                       payload: "0xabcd"
                     )PB"));

  RunProtoPacketTest("IPv6 packet with IPv4 ethertype",
                     gutil::ParseProtoOrDie<Packet>(R"PB(
                       headers {
                         ethernet_header {
                           ethernet_source: "11:22:33:44:55:66"
                           ethernet_destination: "aa:bb:cc:dd:ee:ff"
                           ethertype: "0x0800"
                         }
                       }
                       headers {
                         ipv6_header {
                           version: "0x6"
                           dscp: "0x1b"
                           ecn: "0x1"
                           flow_label: "0x12345"
                           payload_length: "0x0000"
                           next_header: "0x05"
                           hop_limit: "0x10"
                           ipv6_source: "::"
                           ipv6_destination: "f::f"
                         }
                       }
                     )PB"));

  RunProtoPacketTest("IPv6 packet without IPv6 header",
                     gutil::ParseProtoOrDie<Packet>(R"PB(
                       headers {
                         ethernet_header {
                           ethernet_source: "11:22:33:44:55:66"
                           ethernet_destination: "aa:bb:cc:dd:ee:ff"
                           ethertype: "0x86dd"
                         }
                       }
                       payload: "0xabcd"
                     )PB"));
}

}  // namespace packetlib

int main() {
  packetlib::main();
  return 0;
}
