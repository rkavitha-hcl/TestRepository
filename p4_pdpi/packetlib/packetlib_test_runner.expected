================================================================================
Parsing test: Ethernet packet (valid)
================================================================================
-- INPUT -----------------------------------------------------------------------
# ethernet header
    ethernet_source: 0x112233445566
    ethernet_destination: 0xaabbccddeeff
    ether_type: 0x0001  # This means size(payload) = 1 byte.
    # payload
    payload: 0x01
-- OUTPUT ----------------------------------------------------------------------
headers {
  ethernet_header {
    ethernet_source: "11:22:33:44:55:66"
    ethernet_destination: "aa:bb:cc:dd:ee:ff"
    ethertype: "0x0001"
  }
}
payload: "0x01"

================================================================================
Parsing test: Ethernet packet (invalid)
================================================================================
-- INPUT -----------------------------------------------------------------------
# ethernet header
    ethernet_source: 0x112233445566
    ethernet_destination: 0xaabbccddeeff
    ether_type: 0x0001  # This means size(payload) = 1 byte.
    # payload
    pqyload: 0x0102  # 2 bytes, but ether_type says 1 byte.
-- OUTPUT ----------------------------------------------------------------------
headers {
  ethernet_header {
    ethernet_source: "11:22:33:44:55:66"
    ethernet_destination: "aa:bb:cc:dd:ee:ff"
    ethertype: "0x0001"
  }
}
payload: "0x0102"
reasons_invalid: "headers[0].ethertype: value 0x0001 is <= 1500 and should thus match packet size, but packet size is 2"

================================================================================
Parsing test: IPv4 packet (invalid)
================================================================================
-- INPUT -----------------------------------------------------------------------
# ethernet header
    ethernet_source: 0x112233445566
    ethernet_destination: 0xaabbccddeeff
    ether_type: 0x0800
    # IPv4 header:
    version: 0x4
    ihl: 0x5
    dhcp: 0b011011
    ecn: 0b01
    total_length: 0x6fc6
    identification: 0xa3cd
    flags: 0b000
    fragment_offset: 0b0000000000000
    ttl: 0x10
    protocol: 0x05  # some unsupported protocol
    checksum: 0x1234
    ipv4_source: 0x0a000001
    ipv4_destination: 0x14000003
    # other headers:
    payload: 0x1234
-- OUTPUT ----------------------------------------------------------------------
headers {
  ethernet_header {
    ethernet_source: "11:22:33:44:55:66"
    ethernet_destination: "aa:bb:cc:dd:ee:ff"
    ethertype: "0x0800"
  }
}
headers {
  ipv4_header {
    version: "0x4"
    ihl: "0x5"
    dscp: "0x1b"
    ecn: "0x1"
    total_length: "0x6fc6"
    identification: "0xa3cd"
    flags: "0x0"
    fragment_offset: "0x0000"
    ttl: "0x10"
    protocol: "0x05"
    checksum: "0x1234"
    ipv4_source: "10.0.0.1"
    ipv4_destination: "20.0.0.3"
  }
}
payload: "0x1234"
reasons_invalid: "headers[1].total_length: Must be 0x0016, but was 0x6fc6 instead."
reasons_invalid: "headers[1].checksum: Must be 0xaedd, but was 0x1234 instead."

================================================================================
Parsing test: IPv4 packet (valid)
================================================================================
-- INPUT -----------------------------------------------------------------------
# ethernet header
    ethernet_source: 0x112233445566
    ethernet_destination: 0xaabbccddeeff
    ether_type: 0x0800
    # IPv4 header:
    version: 0x4
    ihl: 0x5
    dhcp: 0b011011
    ecn: 0b01
    total_length: 0x0016
    identification: 0xa3cd
    flags: 0b000
    fragment_offset: 0b0000000000000
    ttl: 0x10
    protocol: 0x05  # some unsupported protocol
    checksum: 0xb2e7
    ipv4_source: 0x0a000001
    ipv4_destination: 0x14000003
    # other headers:
    payload: 0x1234
-- OUTPUT ----------------------------------------------------------------------
headers {
  ethernet_header {
    ethernet_source: "11:22:33:44:55:66"
    ethernet_destination: "aa:bb:cc:dd:ee:ff"
    ethertype: "0x0800"
  }
}
headers {
  ipv4_header {
    version: "0x4"
    ihl: "0x5"
    dscp: "0x1b"
    ecn: "0x1"
    total_length: "0x0016"
    identification: "0xa3cd"
    flags: "0x0"
    fragment_offset: "0x0000"
    ttl: "0x10"
    protocol: "0x05"
    checksum: "0xb2e7"
    ipv4_source: "10.0.0.1"
    ipv4_destination: "20.0.0.3"
  }
}
payload: "0x1234"

================================================================================
Parsing test: IPv6 packet (invalid)
================================================================================
-- INPUT -----------------------------------------------------------------------
# ethernet header
    ethernet_source: 0x554433221100
    ethernet_destination: 0xffeeddccbbaa
    ether_type: 0x86DD
    # IPv6 header:
    version: 0x4
    dhcp: 0b011011
    ecn: 0b01
    flow_label: 0x12345
    payload_length: 0x0000
    next_header: 0x90  # some unassigned protocol
    hop_limit: 0xff
    ipv6_source: 0x00001111222233334444555566667777
    ipv6_destination: 0x88889999aaaabbbbccccddddeeeeffff
    # other headers:
    payload: 0x12
-- OUTPUT ----------------------------------------------------------------------
headers {
  ethernet_header {
    ethernet_source: "55:44:33:22:11:00"
    ethernet_destination: "ff:ee:dd:cc:bb:aa"
    ethertype: "0x86dd"
  }
}
headers {
  ipv6_header {
    version: "0x4"
    dscp: "0x1b"
    ecn: "0x1"
    flow_label: "0x12345"
    payload_length: "0x0000"
    next_header: "0x90"
    hop_limit: "0xff"
    ipv6_source: "0:1111:2222:3333:4444:5555:6666:7777"
    ipv6_destination: "8888:9999:aaaa:bbbb:cccc:dddd:eeee:ffff"
  }
}
payload: "0x12"
reasons_invalid: "headers[1].version: Must be 0x6, but was 0x4 instead."
reasons_invalid: "headers[1].payload_length: Must be 0x0001, but was 0x0000 instead."

================================================================================
Parsing test: IPv6 packet (valid)
================================================================================
-- INPUT -----------------------------------------------------------------------
# ethernet header
    ethernet_source: 0x554433221100
    ethernet_destination: 0xffeeddccbbaa
    ether_type: 0x86DD
    # IPv6 header:
    version: 0x6
    dhcp: 0b011011
    ecn: 0b01
    flow_label: 0x12345
    payload_length: 0x0001
    next_header: 0x90  # some unassigned protocol
    hop_limit: 0x03
    ipv6_source: 0x00001111222233334444555566667777
    ipv6_destination: 0x88889999aaaabbbbccccddddeeeeffff
    # other headers:
    payload: 0x12
-- OUTPUT ----------------------------------------------------------------------
headers {
  ethernet_header {
    ethernet_source: "55:44:33:22:11:00"
    ethernet_destination: "ff:ee:dd:cc:bb:aa"
    ethertype: "0x86dd"
  }
}
headers {
  ipv6_header {
    version: "0x6"
    dscp: "0x1b"
    ecn: "0x1"
    flow_label: "0x12345"
    payload_length: "0x0001"
    next_header: "0x90"
    hop_limit: "0x03"
    ipv6_source: "0:1111:2222:3333:4444:5555:6666:7777"
    ipv6_destination: "8888:9999:aaaa:bbbb:cccc:dddd:eeee:ffff"
  }
}
payload: "0x12"

================================================================================
Proto packet test: IPv4 without computed fields
================================================================================
-- INPUT -----------------------------------------------------------------------
packet =
headers {
  ethernet_header {
    ethernet_source: "11:22:33:44:55:66"
    ethernet_destination: "aa:bb:cc:dd:ee:ff"
    ethertype: "0x0800"
  }
}
headers {
  ipv4_header {
    ihl: "0x5"
    dscp: "0x1b"
    ecn: "0x1"
    identification: "0xa3cd"
    flags: "0x0"
    fragment_offset: "0x0000"
    ttl: "0x10"
    protocol: "0x05"
    ipv4_source: "10.0.0.1"
    ipv4_destination: "20.0.0.3"
  }
}
payload: "0xabcd"

-- OUTPUT ----------------------------------------------------------------------
ValidatePacket(packet) = INVALID_ARGUMENT: Packet invalid for the following reasons:
- headers[1].version: missing
- headers[1].total_length: missing
- headers[1].checksum: missing

UpdateComputedFields(packet) = true
packet =
headers {
  ethernet_header {
    ethernet_source: "11:22:33:44:55:66"
    ethernet_destination: "aa:bb:cc:dd:ee:ff"
    ethertype: "0x0800"
  }
}
headers {
  ipv4_header {
    version: "0x4"
    ihl: "0x5"
    dscp: "0x1b"
    ecn: "0x1"
    total_length: "0x0016"
    identification: "0xa3cd"
    flags: "0x0"
    fragment_offset: "0x0000"
    ttl: "0x10"
    protocol: "0x05"
    checksum: "0xb2e7"
    ipv4_source: "10.0.0.1"
    ipv4_destination: "20.0.0.3"
  }
}
payload: "0xabcd"

ValidatePacket(packet) = OK
Serialize(Packet) = OK

================================================================================
Proto packet test: IPv4 with various invalid fields
================================================================================
-- INPUT -----------------------------------------------------------------------
packet =
headers {
  ethernet_header {
    ethernet_destination: "aa:bb:cc:dd:ee:ff"
    ethertype: "0x0800"
  }
}
headers {
  ipv4_header {
    version: "0x3"
    ihl: "0x5k"
    dscp: "0x1b"
    ecn: "0x1"
    identification: "0xa3cd"
    flags: "0x0"
    fragment_offset: "0x0000"
    ttl: "0x10"
    protocol: "0x05"
    ipv4_source: "ffff:1::"
    ipv4_destination: "20.0.0.3"
  }
}
payload: "0xabcd"

-- OUTPUT ----------------------------------------------------------------------
ValidatePacket(packet) = INVALID_ARGUMENT: Packet invalid for the following reasons:
- headers[0].ethernet_source: missing
- headers[1].ihl: invalid format: illegal conversion from hex string 0x5k to 4 bits; expected between 5 and 8 bits
- headers[1].total_length: missing
- headers[1].checksum: missing
- headers[1].ipv4_source: invalid format: Invalid IPv4 address: ffff:1::
- headers[1].version: Must be 0x4, but was 0x3 instead.

UpdateComputedFields(packet) = INVALID_ARGUMENT: illegal conversion from hex string 0x5k to 4 bits; expected between 5 and 8 bits
Serialize(Packet) = INVALID_ARGUMENT: illegal conversion from hex string 0x5k to 4 bits; expected between 5 and 8 bits

================================================================================
Proto packet test: IPv6 without computed fields
================================================================================
-- INPUT -----------------------------------------------------------------------
packet =
headers {
  ethernet_header {
    ethernet_source: "11:22:33:44:55:66"
    ethernet_destination: "aa:bb:cc:dd:ee:ff"
    ethertype: "0x86dd"
  }
}
headers {
  ipv6_header {
    dscp: "0x1b"
    ecn: "0x1"
    flow_label: "0x12345"
    next_header: "0x05"
    hop_limit: "0x10"
    ipv6_source: "::"
    ipv6_destination: "f::f"
  }
}
payload: "0xabcd"

-- OUTPUT ----------------------------------------------------------------------
ValidatePacket(packet) = INVALID_ARGUMENT: Packet invalid for the following reasons:
- headers[1].version: missing
- headers[1].payload_length: missing

UpdateComputedFields(packet) = true
packet =
headers {
  ethernet_header {
    ethernet_source: "11:22:33:44:55:66"
    ethernet_destination: "aa:bb:cc:dd:ee:ff"
    ethertype: "0x86dd"
  }
}
headers {
  ipv6_header {
    version: "0x6"
    dscp: "0x1b"
    ecn: "0x1"
    flow_label: "0x12345"
    payload_length: "0x0002"
    next_header: "0x05"
    hop_limit: "0x10"
    ipv6_source: "::"
    ipv6_destination: "f::f"
  }
}
payload: "0xabcd"

ValidatePacket(packet) = OK
Serialize(Packet) = OK

================================================================================
Proto packet test: IPv6 with various invalid fields
================================================================================
-- INPUT -----------------------------------------------------------------------
packet =
headers {
  ethernet_header {
    ethernet_source: "11:22:33:44:55:66"
    ethernet_destination: "aa:bb:cc:dd:ee:ff"
    ethertype: "0x86dd"
  }
}
headers {
  ipv6_header {
    version: "0x4"
    dscp: "1b"
    ecn: "0b01"
    flow_label: "0x1234"
    payload_length: "0x0000"
    next_header: "0x050"
    hop_limit: "0x1"
    ipv6_source: "20.0.0.3"
    ipv6_destination: ":"
  }
}
payload: "0xabcd"

-- OUTPUT ----------------------------------------------------------------------
ValidatePacket(packet) = INVALID_ARGUMENT: Packet invalid for the following reasons:
- headers[1].dscp: invalid format: illegal conversion from hex string 1b to 6 bits; expected between -3 and 0 bits
- headers[1].ecn: invalid format: illegal conversion from hex string 0b01 to 2 bits; expected between 5 and 8 bits
- headers[1].flow_label: invalid format: illegal conversion from hex string 0x1234 to 20 bits; expected between 13 and 16 bits
- headers[1].next_header: invalid format: illegal conversion from hex string 0x050 to 8 bits; expected between 9 and 12 bits
- headers[1].hop_limit: invalid format: illegal conversion from hex string 0x1 to 8 bits; expected between 1 and 4 bits
- headers[1].ipv6_source: invalid format: On input '20.0.0.3': Invalid IPv6 address: '20.0.0.3'
- headers[1].ipv6_destination: invalid format: On input ':': Invalid IPv6 address: ':'
- headers[1].version: Must be 0x6, but was 0x4 instead.
- headers[1].payload_length: Must be 0x0002, but was 0x0000 instead.

UpdateComputedFields(packet) = false
Serialize(Packet) = INVALID_ARGUMENT: Packet invalid for the following reasons:
- headers[1].dscp: invalid format: illegal conversion from hex string 1b to 6 bits; expected between -3 and 0 bits
- headers[1].ecn: invalid format: illegal conversion from hex string 0b01 to 2 bits; expected between 5 and 8 bits
- headers[1].flow_label: invalid format: illegal conversion from hex string 0x1234 to 20 bits; expected between 13 and 16 bits
- headers[1].next_header: invalid format: illegal conversion from hex string 0x050 to 8 bits; expected between 9 and 12 bits
- headers[1].hop_limit: invalid format: illegal conversion from hex string 0x1 to 8 bits; expected between 1 and 4 bits
- headers[1].ipv6_source: invalid format: On input '20.0.0.3': Invalid IPv6 address: '20.0.0.3'
- headers[1].ipv6_destination: invalid format: On input ':': Invalid IPv6 address: ':'
- headers[1].version: Must be 0x6, but was 0x4 instead.
- headers[1].payload_length: Must be 0x0002, but was 0x0000 instead.

================================================================================
Proto packet test: IPv6 packet with IPv4 ethertype
================================================================================
-- INPUT -----------------------------------------------------------------------
packet =
headers {
  ethernet_header {
    ethernet_source: "11:22:33:44:55:66"
    ethernet_destination: "aa:bb:cc:dd:ee:ff"
    ethertype: "0x0800"
  }
}
headers {
  ipv6_header {
    version: "0x6"
    dscp: "0x1b"
    ecn: "0x1"
    flow_label: "0x12345"
    payload_length: "0x0000"
    next_header: "0x05"
    hop_limit: "0x10"
    ipv6_source: "::"
    ipv6_destination: "f::f"
  }
}

-- OUTPUT ----------------------------------------------------------------------
ValidatePacket(packet) = INVALID_ARGUMENT: Packet invalid for the following reasons:
- headers[1]: expected Ipv4Header, got Ipv6Header

UpdateComputedFields(packet) = false
Serialize(Packet) = INVALID_ARGUMENT: Packet invalid for the following reasons:
- headers[1]: expected Ipv4Header, got Ipv6Header

================================================================================
Proto packet test: IPv6 packet without IPv6 header
================================================================================
-- INPUT -----------------------------------------------------------------------
packet =
headers {
  ethernet_header {
    ethernet_source: "11:22:33:44:55:66"
    ethernet_destination: "aa:bb:cc:dd:ee:ff"
    ethertype: "0x86dd"
  }
}
payload: "0xabcd"

-- OUTPUT ----------------------------------------------------------------------
ValidatePacket(packet) = INVALID_ARGUMENT: Packet invalid for the following reasons:
- headers[1]: header missing - expected Ipv6Header

UpdateComputedFields(packet) = false
Serialize(Packet) = INVALID_ARGUMENT: Packet invalid for the following reasons:
- headers[1]: header missing - expected Ipv6Header

